### PR TITLE
Vocabulary for Tamil Stemmer

### PR DESCRIPTION
**Pull request by rdamodharan**
This pull request has vocabulary and output of the tamil stemming algorithm (ref: https://github.com/snowballstem/snowball/pull/2)

### CODE REVIEW

File: tamil/README.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+These words were taken from Tamil WordNet
+[Tamil WordNet](http://www.au-kbc.org/research_areas/nlp/projects/tamil_wordnet.html)
+
+It can be downloaded from:
+[TamilVU](http://www.tamilvu.org/tsdf/downloads/TamilWordNet.zip) or
+[amrita.edu](http://nlp.amrita.edu:8080/project/mhrd/Wn-t.html)
```
Comment (ojwb) [inline tamil/README.md]: Second link now seems to be: http://nlp.amrita.edu:8080/project/mhrd/tamil_1.html but the file there is much smaller than the one at tamilvu.org (1.1MB vs 58.6MB).

I downloaded http://www.tamilvu.org/tsdf/downloads/TamilWordNet.zip and it tells me that:

```
Tamil Wordnet is Copyright (c) 2005
by AU-KBC Research Centre, Chennai &
Tamil University, Thanjavur, and
licensed under the terms of 
GNU General Public License (GPL)
```

So unfortunately that seems to mean we can't use this as a data source.

**Comment by ojwb** [comment]
I've started to look into this.  So far I've discovered that you only need the preceding line from `voc.txt` to trigger this (so effectively the reproducer is now a two line input file), and that this doesn't trigger errors from valgrind, asan or ubsan.

**Comment by ojwb** [comment]
After further digging, it looks like this is actually a bug in the tamil stemmer implementation - it assumes that `found_vetrumai_urupu` is reset between runs.  I tested with this change which seems to fix the issue (and as far as I can see the code doesn't assume any other variables will be reset):

```
--- a/algorithms/tamil/stem_Unicode.sbl
+++ b/algorithms/tamil/stem_Unicode.sbl
@@ -435,6 +435,7 @@ define remove_tense_suffix as (
 )
 
 define stem as (
+ unset found_vetrumai_urupu
  do fix_ending
  has_min_length
  do remove_question_prefixes
```

With this patch, there are 987 words in `tamil/voc.txt` which now don't produce the stem listed in `tamil/output.txt`. @rdamodharan - can you review these and let us know if they are OK?

**Comment by rdamodharan** [comment]
@ojwb Thank you for looking into this. I will review this and update the ticket.

**Comment by rdamodharan** [comment]
@ojwb, I did some more testing and I think that the change seems fine.

**Comment by ojwb** [comment]
On reflection, having test data under different licences such as the GPL isn't really a big deal, so I've just been looking at the licensing situation for this data some more.

As I noted, http://www.tamilvu.org/tsdf/downloads/TamilWordNet.zip is GPL, but that doesn't actually seem to contain the wordnet data.  It's mostly software, and the only data I can find in it is an SQL database dump, which all seems to be transliterated into the Latin alphabet.

The 1.1MB file from http://nlp.amrita.edu:8080/project/mhrd/tamil_1.html does seem to contain data, but there's no licensing information I can find in the download or anywhere on the site.  That means we just can't use the data.

@rdamodharan Can you clarify which data source(s) you used to generate the `voc.txt` file in this PR, and how you turned them into `voc.txt`?

If this source isn't suitable, we could take a dump from Tamil wikipedia, and produce a word list from that.  The source data is then CC-BY-SA 3.0, and so we can safely use any derived form under that licence too.

**Comment by rdamodharan** [comment]
@ojwb, I used the Tamil WordNet data for voc.txt - https://github.com/rdamodharan/tamil-stemmer/blob/master/docs/Evaluation.md. I loaded the SQL dump into MySQL database and used a perl script to convert the data to utf-8 encoded data. It looks like I didn't check in the perl script used to convert the data. I will see if I can find it out and add to the repository.

**Comment by ojwb** [comment]
Great, so the source is GPL, and we can use this.  I'll sort that out.

It would be useful to have the script available in case there's newer data to convert, etc.

BTW, you have a vim swap file checked into that repo: `.tamil-stem_Unicode.sbl.swp`

**Comment by rdamodharan** [comment]
Thanks Olly. I am searching through my backups to find the script. I will
send in a pull request once I get the script to transliterate the data.

Damu

**Comment by ojwb** [comment]
I realised that it's problematic for the original source to be GPL when there's no means provided to transform from the original.

But having no Tamil test data is also problematic, as we'll miss any changes to the snowball compiler or the Tamil snowball stemmer code which change the behaviour of the stemmer.

So I've added a Tamil wordlist extracted from a Tamil Wikipedia dump, along with the simple script used to generate it (which essentially just collates matches for the regexp `\b[\x{b80}-\x{bff}]+\b` and then outputs any which occur 3 or more times).  This contains 443271 unique words.

Improvements are certainly welcome.

**Comment by rdamodharan** [comment]
Hi Oliv,
Thanks for closing it out with new data set. I apologize for not getting
this fixed earlier.

Damu

**Pull request closed without merging**